### PR TITLE
Update alephant preview to serve batched components

### PR DESCRIPTION
**Pull request by samfrench**
## Problem

Alephant Preview only serves individual components. To test polling locally, we need to serve batch components.

## Solution

Add batch post endpoint to Alephant Preview.

### CODE REVIEW

**Comment by Integralist** [comment]
@samfrench nothing too crazy in this PR as far as I can tell. I presume you've tested this successfully with both single and batch components? If you could confirm please so I can give appropriate thumbs up

**Comment by samfrench** [comment]
@Integralist I have tested the batch endpoint and the existing single component loading endpoint and they both work. I haven't merged the other PR yet (https://github.com/BBC-News/alephant-preview/pull/12) which fixes the fixture loading, but hoping to roll out all changes together when releasing the gem.

**Comment by samfrench** [comment]
cc @leostaples

**Comment by Integralist** [comment]
@samfrench all good then, as long as you're confident everything is working I can serve you up a slice of :cake:

**Comment by stevenjack** [comment]
@Integralist @samfrench maybe a test in https://github.com/BBC-News/alephant-preview/blob/master/spec/integration/preview_spec.rb would sure things up in terms of knowing it's not breaking other things?

**Comment by samfrench** [comment]
Thanks @stevenjack, I've added a post spec now.

@Integralist, can I get the :+1: on this please?

File: spec/integration/preview_spec.rb [hunk 1]
```diff
@@ -59,6 +59,44 @@
     end
   end
 
+  describe "component batch endpoint (POST /components/batch" do
+
+    describe "content" do
+      before(:each) do
+        post "/components/batch", { :components => [ { :component => id, :options => { :fixture => id } } ] }.to_json
```
Comment (Integralist) [inline spec/integration/preview_spec.rb]: Could we multi-line this hash please

File: spec/integration/preview_spec.rb [hunk 2]
```diff
@@ -59,6 +59,44 @@
...
+      end
+
+      let (:response) { JSON.parse(last_response.body.chomp, :symbolize_names => true) }
+
+      context "without a data mapper" do
+        let(:id) { "foo" }
+
+        expected = { :components => [ { :component => "foo", :options => {}, :status => 200, :body => "content\n" } ] }
```
Comment (Integralist) [inline spec/integration/preview_spec.rb]: Same again, can we multi-line these please as I'd prefer it to be under 80 characters

File: spec/integration/preview_spec.rb [hunk 3]
```diff
@@ -59,6 +59,44 @@
...
+
+        specify { expect(response).to eq(expected) }
+      end
+
+      context "with a data mapper" do
+
+        context "using a single fixture" do
+          let (:id) { "bar" }
+
+          expected = { :components => [ { :component => "bar", :options => {}, :status => 200, :body => "data mapped content\n" } ] }
```
Comment (Integralist) [inline spec/integration/preview_spec.rb]: Multi-line

File: spec/integration/preview_spec.rb [hunk 4]
```diff
@@ -59,6 +59,44 @@
...
+
+          specify { expect(response).to eq(expected) }
+        end
+
+        context "using multiple fixtures" do
+          let (:id) { "baz" }
+
+          expected = { :components => [ { :component => "baz", :options => {}, :status => 200, :body => "multiple endpoint data mapped content\n" } ] }
```
Comment (Integralist) [inline spec/integration/preview_spec.rb]: Multi-line

**Comment by BBCMarkMcDonnell** [comment]
@samfrench when I run `bundle exec rspec` I get a passing suite of tests but I've noticed the following error...

```
...WARNING: Using the `raise_error` matcher without providing a specific error or message risks false positives, since `raise_error` will match when Ruby raises a `NoMethodError`, `NameError` or `ArgumentError`, potentially allowing the expectation to pass without even executing the method you are intending to call. Actual error raised was #<RuntimeError: There isn't a fixture matching the request call, please add one>. Instead consider providing a specific error class or message. This message can be supressed by setting: `RSpec::Expectations.configuration.warn_about_potential_false_positives = false`. Called from /Users/M/Projects/BBC/alephant-preview/spec/fixture_loader_spec.rb:44:in `(root)'.
.........

Finished in 0.332 seconds (files took 0.915 seconds to load)
12 examples, 0 failures
```

I got this same message myself when doing a pull request for some other Alephant gems. I feel it's worth you fixing up the code to resolve this error (here was my fix, so you have an idea... https://github.com/BBC-News/alephant-renderer/commit/862d0046338cb7a18ce3c32bff90f607736021dc)

**Comment by samfrench** [comment]
@Integralist, I have changed the expected fixtures to all be under 80 characters, and provided an alternative to fix the error warning.

**Comment by Integralist** [comment]
Morning @samfrench great stuff. But it looks like the Travis CI build is breaking

**Comment by samfrench** [comment]
@Integralist, this one also passed on Travis CI now.

File: lib/alephant/preview/server.rb [hunk 5]
```diff
@@ -27,6 +27,10 @@ class Server < Sinatra::Base
 
       BASE_LOCATION = "#{(ENV['BASE_LOCATION'] || Dir.pwd)}/components"
 
+      before do
+        response["Access-Control-Allow-Origin"] = "*"
```
Comment (BBCMarkMcDonnell) [inline lib/alephant/preview/server.rb]: @samfrench out of interest, why's this needed?

File: lib/alephant/preview/server.rb [hunk 6]
```diff
@@ -41,6 +45,22 @@ class Server < Sinatra::Base
         render_component
       end
 
+      post "/components/batch" do
+        batch_components = []
+
+        ((JSON.parse(request.body.read, :symbolize_names => true) || {}).fetch(:components, [])).each do |component|
```
Comment (BBCMarkMcDonnell) [inline lib/alephant/preview/server.rb]: @samfrench could we break this line out into multiple method calls please. 

e.g. something like...

```ruby
post "/components/batch" do
  ...

  batched_components.each do |component|
    # stuff
  end

  ...
end

def batched_components
  request_body.fetch(:components, [])
end

def request_body
  JSON.parse(request.body.read, :symbolize_names => true) || {}
end
```

File: lib/alephant/preview/server.rb [hunk 7]
```diff
@@ -41,6 +45,22 @@ class Server < Sinatra::Base
...
+          options = component.fetch(:options, {}) || {}
+          params["template"] = component.fetch(:component)
+          params["id"] = find_id_from_template params["template"]
+          params["fixture"] = options.fetch(:fixture, "responsive") || "responsive"
+          batch_components << render_batch_component
+        end
+
+        {
```
Comment (BBCMarkMcDonnell) [inline lib/alephant/preview/server.rb]: I'd probably single line this...

```rb
{ :components => batch_components }.to_json
```

File: lib/alephant/preview/server.rb [hunk 8]
```diff
@@ -63,6 +83,15 @@ def render_component
         view_mapper.generate(fixture_data)[template].render
       end
 
+      def render_batch_component
+        {
+          :component => template,
+          :options => {},
+          :status => 200,
+          :body => render_component
```
Comment (BBCMarkMcDonnell) [inline lib/alephant/preview/server.rb]: Could we get the `=>` aligned

File: spec/fixture_loader_spec.rb [hunk 9]
```diff
@@ -41,7 +41,8 @@
       context 'using incorrect amount of fixtures' do
         it "should raise an exception" do
           (0..2).each { |index| subject.get(uri).body }
-          expect{ subject.get(uri).body }.to raise_error
+          expect{ subject.get(uri).body }.to raise_error(RuntimeError,
```
Comment (BBCMarkMcDonnell) [inline spec/fixture_loader_spec.rb]: I'd probably write this like so...

```ruby
expect do 
  subject.get(uri).body
end.to raise_error(
  RuntimeError, "There isn't a fixture matching the request call, please add one"
)
```

File: spec/integration/preview_spec.rb [hunk 10]
```diff
@@ -59,6 +59,80 @@
...
+        }.to_json
+      end
+
+      let (:response) { JSON.parse(last_response.body.chomp, :symbolize_names => true) }
+
+      context "without a data mapper" do
+        let(:id) { "foo" }
+
+        expected = {
+          :components => [
```
Comment (BBCMarkMcDonnell) [inline spec/integration/preview_spec.rb]: @samfrench Same alignment comment as above

Comment (BBCMarkMcDonnell) [inline spec/integration/preview_spec.rb]: Same for all of below :-)

File: lib/alephant/preview/server.rb (hunk 5 above)
Comment (samfrench) [inline lib/alephant/preview/server.rb]: We are making a JS post request in our browser for the components. I get a cross access error if this isn't present.

**Comment by samfrench** [comment]
@Integralist, I have addressed all of your comments for this PR.

File: lib/alephant/preview/server.rb (hunk 5 above)
Comment (BBCMarkMcDonnell) [inline lib/alephant/preview/server.rb]: OK cool, that makes sense :+1:

**Comment by BBCMarkMcDonnell** [comment]
@samfrench :+1: / :cake:

**Pull request merged**